lib/Sema/MiscDiagnostics.cpp

```diff
@@ -1444,6 +1444,68 @@ static void diagRecursivePropertyAccess(const Expr *E, const DeclContext *DC) {
   const_cast<Expr *>(E)->walk(walker);
 }
 
+static void
+diagnoseMemberMethodAssignmentsCapturingSelf(const Expr *E,
+                                             const DeclContext *DC) {
+
+  class DiagnoseWalker : public ASTWalker {
+    ASTContext &Ctx;
+
+  public:
+    explicit DiagnoseWalker(ASTContext &ctx) : Ctx(ctx) {}
+
+    bool walkToDeclPre(Decl *D) override { return false; }
+
+    bool shouldWalkIntoSeparatelyCheckedClosure(ClosureExpr *expr) override {
+      return false;
+    }
+
+    bool shouldWalkCaptureInitializerExpressions() override { return true; }
+
+    bool shouldWalkIntoTapExpression() override { return false; }
+
+    DeclRefExpr *getMemberFuncReferencingSelf(Expr *selfParam,
+                                              AutoClosureExpr *ACE) {
+
+      if (auto *ICE = dyn_cast<AutoClosureExpr>(ACE->getSingleExpressionBody()))
+        if (auto *AE = dyn_cast<ApplyExpr>(ICE->getSingleExpressionBody()))
+          if (auto *DSCE = dyn_cast<DotSyntaxCallExpr>(AE->getFn()))
+            if (auto *base = dyn_cast<DeclRefExpr>(DSCE->getBase()))
+              if (base->getType()->getCanonicalType() ==
```

Comment:
This is too broad because you're also going to warn in a capture of any value with the same type as 'self'